batch.py

```diff
@@ -0,0 +1,761 @@
+- run association tests
+"""
+
+# import python modules # is there a specific order, rationale for grouping imports??
+import os
+
+import click
+import logging
+
+from cloudpathlib import AnyPath
+from cpg_utils import to_path
+from cpg_utils.hail_batch import (
+    dataset_path,
+    get_config,
+    init_batch,
+    output_path,
+    remote_tmpdir,
+)
+
+import numpy as np
+import pandas as pd
+import xarray as xr  
+
+from limix.qc import quantile_gaussianize
+from numpy import eye, ones
+from pandas_plink import read_plink1_bin
+from scipy.stats import shapiro
+
+# qvalue imports
+import scipy as sp
+from scipy import interpolate
+
+import hail as hl
+import hailtop.batch as hb
+from hail.methods import export_plink
+
+from cellregmap import (  # figure out how to import this from github
+    run_gene_set_association,
+    run_burden_association,
+    omnibus_set_association,
+)
+
+# use logging to print statements, display at info level
+logging.basicConfig(
+    format="%(levelname)s:%(message)s", level=logging.INFO
+)  # consider removing some print statements?
+
+DEFAULT_JOINT_CALL_MT = dataset_path("mt/v7.mt")
+DEFAULT_ANNOTATION_HT = dataset_path("tob_wgs_vep/104/vep104.3_GRCh38.ht")  # atm VEP only
+
+# region SUBSET_VARIANTS
+
+
+def filter_variants(
+    mt_path: str,
+    samples: list[str],
+    output_mt_path: str,
+):
+    """Subset hail matrix table
+
+    Input:
+    joint call hail matrix table
+    set of samples for which we have scRNA-seq data
+
+    Output:
+    subsetted hail matrix table, containing only variants that:
+    1) are not ref-only, 2) biallelic, 3) meet QC filters, 4) are rare (MAF<5%)
+    """
+    # read hail matrix table object (WGS data)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # subset to relevant samples (samples we have scRNA-seq data for)
+    mt = mt.filter_cols(samples)  # figure out syntax
+
+    # densify (can this be done at the end?)
+    mt = hl.experimental.densify(
+        mt
+    )  # is this definitely efficient to do on the whole thing once, vs on much smaller subset many times?
+
+    # filter out low quality variants and consider biallelic variants only (no multi-allelic, no ref-only)
+    mt = mt.filter_rows(  # check these filters!
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)  # QC
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt = mt.checkpoint(output_mt_path, overwrite=True)  # syntax???
+    logging.info(
+        f"Number of rare (freq<5%) and QC'd biallelic variants: {mt.count()[0]}"
+    )
+
+    return output_mt_path  # both input and output??
```

Comment:
if this is passed as an input you don't need to return it (unless you are altering it in some way, which is not being done here)

Reply:
right so then that path can just be passed to the next job, and all this function is doing is writing to that path? how do i check that has been run before i pass this path to the next thing